### ISSUE TITLE
Ideas from the magpie language

### ISSUE THREAD

**Issue by odino** [abs-lang/abs#177]
I found @haifenghuang's magpie and I think there are some useful ideas we could borrow :)

* [const](https://github.com/haifenghuang/magpie/commit/56c9d8599dd0e429cbc901b3e7062baca55a9a3a)
* [exceptions](https://github.com/haifenghuang/magpie#exception-handlingtry-catch-finally)
* [regexes](https://github.com/haifenghuang/magpie#regular-expression)
* [default values for function arguments](https://github.com/haifenghuang/magpie#function)

Regexes are kind of my favorite -- we could build on top on abs's similarity operator to do something like `string ~ /pattern/`. Exceptions are also extra-useful.

**Comment by odino** [comment]
Very keen on regexes and exceptions :)

**Comment by ntwrick** [comment]
Agree. Maybe `"abc" ~ '*c'` works too? Do we really need special regex
pattern delimiters? Any string to right of tilde may contain regex?

**Comment by ntwrick** [comment]
Ignore the back ticks...

**Comment by odino** [comment]
Though that will break the current functionality which is to do case-insensitive string comparison

**Comment by skx** [comment]
Regexps are very useful, and weren't too hard for me to implement in my own monkey-inspired language:

https://github.com/skx/monkey/pull/54

That said default arguments turned out to be much more useful, and [almost trivial](https://github.com/skx/monkey/commit/aba67e65e2e9f73375dcc9e466a82cf86d3e1ef1) to implement.

(Of course your own code is different, but figured references might not hurt!)

**Comment by odino** [comment]
Thanks @skx! Options params made it in https://github.com/abs-lang/abs/pull/369 but I'm still on the fence on regexes. Mostly because I used the ~ operator for case insensitive string comparison so I've just been delaying a decision on what operator to use in regexes (~ or ~=)
